Forwarded: https://github.com/BuddiesOfBudgie/budgie-desktop/pull/329
Author: Jacob-Vlijm <vlijm@planet.nl>
Last-Update: 2023-03-14
Description: [PATCH] Fix panel growing memory consumption over increasing numbers
 of notifications.

---
 src/raven/notifications_group.vala | 14 ++++++++------
 src/raven/notifications_view.vala  | 13 +++++++------
 2 files changed, 15 insertions(+), 12 deletions(-)

diff --git a/src/raven/notifications_group.vala b/src/raven/notifications_group.vala
index df16dd39..e2454b48 100644
--- a/src/raven/notifications_group.vala
+++ b/src/raven/notifications_group.vala
@@ -113,9 +113,10 @@ namespace Budgie {
 		 * dismiss_all is responsible for dismissing all notifications
 		 */
 		public void dismiss_all() {
-			notifications.foreach_steal((id, notification) => {
-				list.remove(notification.get_parent());
-				notification.destroy();
+			notifications.foreach_remove((id, notification) => {
+				var parent = notification.get_parent();
+				list.remove(parent);
+				parent.destroy();
 				dismissed_notification(id);
 				return true;
 			});
@@ -131,10 +132,11 @@ namespace Budgie {
 			var notification = notifications.lookup(id); // Get our notification
 
 			if (notification != null) { // If this notification exists
-				notifications.steal(id);
-				list.remove(notification.get_parent());
+				notifications.remove(id);
+				var parent = notification.get_parent();
+				list.remove(parent);
 				list.invalidate_sort();
-				notification.destroy(); // Nuke the notification
+				parent.destroy();
 				update_count(); // Update our counter
 				dismissed_notification(id); // Notify anything listening
 				if (count == 0) { // This was the last notification
diff --git a/src/raven/notifications_view.vala b/src/raven/notifications_view.vala
index 2b960d8c..9b6e1ed2 100644
--- a/src/raven/notifications_view.vala
+++ b/src/raven/notifications_view.vala
@@ -292,14 +292,16 @@ namespace Budgie {
 						this.listbox.add(group);
 
 						group.dismissed_group.connect((name) => { // When we dismiss the group
-							listbox.remove(group.get_parent()); // Remove this from the listbox
+							var parent = group.get_parent();
+							listbox.remove(parent);
+							parent.destroy();
 
 							/**
-							* If we're not performing a clear all, steal this entry from notifications list and update our child count
-							* Performing a steal seems to affect a .foreach call, so best to avoid this.
+							* If we're not performing a clear all, remove this entry from notifications list and update our child count
+							* Performing a remove seems to affect a .foreach call, so best to avoid this.
 							*/
 							if (!performing_clear_all) {
-								notification_groups.steal(name); // Remove notifications group from list
+								notification_groups.remove(name);
 								update_child_count();
 							}
 
@@ -388,8 +390,7 @@ namespace Budgie {
 				notification_group.dismiss_all();
 			});
 
-			notification_groups.steal_all(); // Ensure we're resetting notifications_list
-
+			notification_groups.remove_all(); // Ensure we're resetting notifications_list
 			performing_clear_all = false;
 			update_child_count();
 			Raven.get_instance().ReadNotifications();
-- 
2.34.1

